WalletWasabi/Extensions/NBitcoinExtensions.cs

```diff
@@ -134,6 +135,36 @@ public static int GetAnonymitySet(this Transaction me, int outputIndex)
 
 		public static int GetAnonymitySet(this Transaction me, uint outputIndex) => GetAnonymitySet(me, (int)outputIndex);
 
+		public static int GetAnonymitySet(this Transaction me, int outputIndex, IEnumerable<SmartCoin> spentOwnCoins)
+		{
+			// Get the anonymity set of i-th output in the transaction.
+			var anonset = me.GetAnonymitySet(outputIndex);
+			var numberOfOwnInputs = spentOwnCoins.Count();
+			// If we provided inputs to the transaction.
+			if (numberOfOwnInputs > 0)
+			{
+				// Take the input that we provided with the smallest anonset.
+				// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
+				var smallestInputAnon = spentOwnCoins.Min(x => x.AnonymitySet);
+
+				// Punish consilidation exponenetially.
+				// If there is only a single input then the exponent should be zero to divide by 1 thus retain the input coin anonset.
+				var consolidatePenalty = Math.Pow(2, numberOfOwnInputs - 1);
```

Comment:
I don't know if this is also considered more idiomatic in C# but since it's a power of two bit shifting will do:

```suggestion
				var consolidatePenalty = 1 << (numberOfOwnInputs - 1);
```

Reply:
We prefer readability.

Reply:
actually readability was my main consideration but i can see how the argument goes both ways ;-)